### PR TITLE
add support for Ecto 3

### PR DESCRIPTION
**Pull request by wiserfirst**
This is an attempt to support Ecto 3, so that deprecation warning for loggers section is fixed (Issue #8).
Thanks @novaugust for [the idea](https://github.com/spandex-project/spandex_ecto/issues/8#issuecomment-478376776)

### CODE REVIEW

**Comment by zachdaniel** [comment]
So this is a good start, but the only problem is that these traces won't be linked to the parent span that started the trace, correct? Don't telemetry handlers happen out of band in a separate process?

**Comment by novaugust** [comment]
@zachdaniel negative, telemetry happens in the process itself
[source](https://github.com/beam-telemetry/telemetry/blob/master/src/telemetry.erl#L126-L142)

**Comment by novaugust** [comment]
I should add though, ecto had a breaking change on telemetry between 3.0 and 3.1, so not only do you have to add telemetry, but you have to add it twice

File: lib/spandex_ecto/telemetry_adapter.ex [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  # This is for ecto_sql >= 3.1
+  def handle_event([_app_name, repo_name, :query], measurements, metadata, _config) when is_map(measurements) do
+    log_entry = %{
+      query: metadata.query,
+      source: metadata.source,
+      params: metadata.params,
+      query_time: Map.get(measurements, :query_time, 0),
+      decode_time: Map.get(measurements, :decode_time, 0),
+      queue_time: Map.get(measurements, :queue_time, 0),
+      result: {metadata.result, "n/a"}
```
Comment (novaugust) [inline lib/spandex_ecto/telemetry_adapter.ex]: I think they're passing the correct values again in `metadata.result` after this revert:
https://github.com/elixir-ecto/ecto_sql/commit/11b83c58b97c1d4a7f0a2a882f048929f294eee6

Comment (wiserfirst) [reply lib/spandex_ecto/telemetry_adapter.ex]: Thanks again @novaugust. From the commit you mentioned, it looks like to me `metadata.result` used to be an `:ok` or `:error` atom, which is why you wrapped it in a tuple; but now it's a tuple already, so wrapping is not necessary anymore?

Comment (novaugust) [reply lib/spandex_ecto/telemetry_adapter.ex]: Right. I was compensating for their breaking change in ecto 3.1.0 - they patched it in 3.1.1, see their changelog for the details =)

Comment (wiserfirst) [reply lib/spandex_ecto/telemetry_adapter.ex]: @novaugust Thanks for the confirmation. I've added a new commit to handle both atom and tuple for metadata.result

**Comment by wiserfirst** [comment]
@zachdaniel I hope @novaugust's comment has answered your question?

Please lemme know if you notice anything wrong or got more suggestions. Thanks!

**Comment by zachdaniel** [comment]
@wiserfirst yes it does thanks! I’ll try and get this merged and released tomorrow morning :)

**Comment by novaugust** [comment]
> @wiserfirst yes it does thanks! I’ll try and get this merged and released tomorrow morning :)

Might double check that this actually works too ^.^ My spandex has been down for a few weeks and I haven't had the chance to look at why.

@wiserfirst are you using this code in production? which version of ecto_sql?

**Comment by zachdaniel** [comment]
I didn’t get a chance to look at this today. I also don’t have a production level service reporting to datadog anymore, which makes this kind of a hassle. @gregmefford, is your example app still a thing we could use to test this?

**Comment by GregMefford** [comment]
Yes, `spandex_example` should be able to check this (and I can get you the API key if you need it). Ive unfortunately been pretty swamped lately and haven’t had as much time as I’d like to get PRs rested and merged.

**Comment by wiserfirst** [comment]
@zachdaniel I used very similar code in prod (basically `TelemetryAdapter` in our code), but not this exact code.

@GregMefford can you point me to more info about how to use `spandex_example` for testing and maybe send me the API key as well? I'm happy to run necessary tests and verify this works in prod. Thanks!

File: README.md [hunk 2]
```diff
@@ -28,16 +28,34 @@ def deps do
...
 config :my_app, MyApp.Repo,
   loggers: [{Ecto.LogEntry, :log, [:info]}, {SpandexEcto.EctoLogger, :trace, ["database_name"]}]
 
 ```
+
+### For Ecto 3
+
+```elixir
+# in application.ex
+:telemetry.attach("spandex-query-tracer", [:my_app, :repo_name, :query], &SpandexEcto.TelemetryAdapter.handle_event/4)
```
Comment (khelll) [inline README.md]: It seems there is [only](https://hexdocs.pm/telemetry/telemetry.html#attach-4) `attach/4`?

I had to do:

```elixir
:telemetry.attach("spandex-query-tracer", [:my_app, :repo_name, :query], &SpandexEcto.TelemetryAdapter.handle_event/4, nil)
```

Comment (wiserfirst) [reply README.md]: You are totally right. Sorry for the miss.

**Pull request merged**